pkg/gateway/services/stub.go

```diff
@@ -106,6 +76,42 @@ func (gws *GatewayService) GetOrCreateStub(ctx context.Context, in *pb.GetOrCrea
 		Extra:              json.RawMessage(in.Extra),
 	}
 
+	// Ensure GPU count is at least 1 if a GPU is required
+	if stubConfig.RequiresGPU() && in.GpuCount == 0 {
+		stubConfig.Runtime.GpuCount = 1
+	}
+
+	if stubConfig.RequiresGPU() {
+		concurrencyLimit, err := gws.backendRepo.GetConcurrencyLimitByWorkspaceId(ctx, authInfo.Workspace.ExternalId)
+		if err != nil && concurrencyLimit != nil && concurrencyLimit.GPULimit <= 0 {
+			return &pb.GetOrCreateStubResponse{
+				Ok:     false,
+				ErrMsg: "GPU concurrency limit is 0.",
+			}, nil
+		}
+
+		gpuCounts, err := gws.providerRepo.GetGPUCounts(gws.appConfig.Worker.Pools)
+		if err != nil {
+			return &pb.GetOrCreateStubResponse{
+				Ok:     false,
+				ErrMsg: "Failed to get GPU counts.",
+			}, nil
+		}
+
+		// T4s are currently in a different pool than other GPUs and won't show up in gpu counts
+		lowGpus := []string{}
+
+		for _, gpu := range gpus {
+			if gpuCounts[gpu.String()] <= 1 && gpu.String() != types.GPU_T4.String() {
```

Comment:
I know this is probably out of scope, but should we remove this T4 check now that we have external provider T4 availability?

Reply:
That is true, but why would we have different behavior for A10Gs and T4s? A10Gs can also be scaled out in AWS since we have them as a lower priority worker pool.

Reply:
We can work on this another time since its out of scope anyway.